src/cc/bcc_syms.cc

```diff
@@ -100,12 +100,38 @@ void ProcSyms::refresh() {
 
 int ProcSyms::_add_module(const char *modname, uint64_t start, uint64_t end,
                           void *payload) {
+  struct ns_cookie nsc = {-1, -1};
+  bool ns_switch = false;
+  int arc;
   ProcSyms *ps = static_cast<ProcSyms *>(payload);
   auto it = std::find_if(ps->modules_.begin(), ps->modules_.end(),
                [=](const ProcSyms::Module &m) { return m.name_ == modname; });
-  if (it == ps->modules_.end())
-    it = ps->modules_.insert(ps->modules_.end(), modname);
+  if (it == ps->modules_.end()) {
+    // If modname references a perf-map, determine if we need to enter a mount
+    // namespace in order to read symbols from it later.
+    if (strstr(modname, ".map") != nullptr) {
+      ns_switch = bcc_procutils_enter_mountns(ps->pid_, &nsc);
+      if (ns_switch) {
+        char new_modname[4096];
+        arc = access(modname, R_OK);
+        bcc_procutils_exit_mountns(&nsc);
+
+        if (arc != 0) {
+          snprintf(new_modname, sizeof (new_modname), "/tmp/perf-%d.map",
```

Comment:
Sorry for the slow reply.  Not 100% sure I understand the question.   I believe this code was attempting to catch the case where the perf map was in the host system instead of a container, but the process that it referred to was inside a different PID namespace.  In that case, we were attempting to resolve the perf map relative to the primordial mount namespace.

To make things more confusing, there's no canonical definition of what has to be in a linux container, so it's possible to get a process that's in a different mount namespace but shares a pid namespace with the host.  Similarly, it's also possible that a process could be in a different pid namespace but share a mount namespace.  Or they both could be disjoint.  You're correct that chroot isn't out of scope after entering a mountns, but it is awfully redundant for a process to create a full chroot inside a container, which if properly scoped should be its own chroot on steroids.  If you really wanted to be thorough, you could attempt to look up the file relative to the mount namespace, and then resolved against any chroot applied by the process after it entered the mount namespace.  But again, I'm not quite sure I understand what's being asked.

The code above is just trying to locate a perf-map using a PID that would resolve on the host system.